bin/reth/src/args/utils.rs

```diff
@@ -106,18 +121,126 @@ pub fn parse_socket_address(value: &str) -> eyre::Result<SocketAddr, SocketAddre
+        "istanbulBlock": 0,
+        "berlinBlock": 0,
+        "londonBlock": 0,
+        "terminalTotalDifficulty": 0,
+        "terminalTotalDifficultyPassed": true,
+        "shanghaiTime": 0
+    }
+}
+"#;
+
+        println!("raw: {:?}", custom_genesis_from_json);
+        println!("this should be true then: {:?}", custom_genesis_from_json.starts_with("\n"));
+        let chain_from_json = genesis_value_parser(&custom_genesis_from_json).unwrap();
+
+        // using structs
+        let config = ChainConfig {
+            chain_id: 2600,
+            homestead_block: Some(0),
+            eip150_block: Some(0),
+            eip155_block: Some(0),
+            eip158_block: Some(0),
+            byzantium_block: Some(0),
+            constantinople_block: Some(0),
+            petersburg_block: Some(0),
+            istanbul_block: Some(0),
+            berlin_block: Some(0),
+            london_block: Some(0),
+            shanghai_time: Some(0),
+            terminal_total_difficulty: Some(U256::ZERO),
+            terminal_total_difficulty_passed: true,
+            ..Default::default()
+        };
+        let genesis = Genesis {
+            config,
+            nonce: 0,
+            timestamp: 1698688670,
+            gas_limit: 5000,
+            difficulty: U256::ZERO,
+            mix_hash: B256::ZERO,
+            coinbase: Address::ZERO,
+            ..Default::default()
+        };
+
+        // seed accounts after genesis struct created
+        let address = hex!("6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b").into();
+        let account = GenesisAccount::default().with_balance(U256::from(33));
+        let genesis = genesis.extend_accounts(HashMap::from(
+            [(address, account)]
+        ));
+
+        println!("\n\ngenesis: {genesis:?}\n\n");
+
+        let custom_genesis_from_struct = serde_json::to_string(&genesis).unwrap();
+        let chain_from_struct = genesis_value_parser(&custom_genesis_from_struct).unwrap();
+        assert_eq!(chain_from_json.genesis(), chain_from_struct.genesis());
+
+        // chain spec
+        let chain_spec = ChainSpecBuilder::default()
+            .chain(2600.into())
+            .genesis(genesis)
+            .cancun_activated()
+            .build();
+
+        println!("\n\nchain_spec: {chain_spec:?}\n\n");
+
+        let chain_spec_json = serde_json::to_string(&chain_spec).unwrap();
+        println!("\n\nchain_spec_json: {chain_spec_json:?}\n\n");
+        let custom_genesis_from_spec = genesis_value_parser(&chain_spec_json).unwrap();
+
+        println!("custom_genesis_from_spec: {:?}", custom_genesis_from_spec.chain());
+        println!("chain_from_struct: {:?}", chain_from_struct.chain());
+        println!("\n\nchain_from_struct: {chain_from_struct:?}\n\n");
+        println!("\n\ncustom_genesis_from_spec {custom_genesis_from_spec:?}\n\n");
+
+        // TODO: currently failing
+        assert_eq!(custom_genesis_from_spec.chain(), chain_from_struct.chain());
+    }
```

Comment:
I expect this to work, but it doesn't. I'm leaving debug print statements in the PR for now and taking a break.

Reply:
Specifically, the chain id from the `genesis_value_parser` has `Chain::Id(0)` instead of the expected `Chain::Id(2600)`

Reply:
this is like an issue with the Serialize  of the chainId, it should be serialized as u64, which needs to be fixed

Reply:
I confirmed the ChainId is serialized/deserialized correctly, and the ChainSpec is serialized correctly. However, when the ChainSpec is deserialized into `AllGenesisFormats` it is created as the `Geth(Genesis)` variant, instead of `Reth(ChainSpec)`

Reply:
Since `AllGenesisFormats` is `#[serde(untagged)]` it just matches to the first enum that deserializes. Changing the order in the enum works, but I'm not sure that's the best solution here

Reply:
please advise @Rjected

Reply:
After further testing, I'm confident this is the correct solution